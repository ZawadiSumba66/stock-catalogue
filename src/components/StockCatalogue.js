import { connect } from 'react-redux';
import { useEffect } from 'react';
import PropTypes from 'prop-types';
import { getstock } from '../actions/index';
import StockItem from './StockItem';
import CategoryFilter from './CategoryFilter';
import filterStock from '../actions/filter';
import store from '../store';

const StockCatalogue = ({
  getStock, loading, stock, filter,
}) => {
  useEffect(() => {
    getStock();
  }, []);

  const handleFilterChange = (e) => {
    store.dispatch(filterStock(e.target.value));
  };

  if (loading) {
    return <p>Please wait...</p>;
  }

  const gainer = filter === 'Gainers' ? '+' : '-';
  let filteredStock = stock;

  if (filter !== 'ALL') {
    filteredStock = stock.filter((gain) => gain.changesPercentage.includes(gainer));
  }
  return (
    <div>
      <div className="header">
        <span className="logo">
          <i className="fas fa-poll" />
          STOCK MARKET
        </span>
        <CategoryFilter handleFilterChange={handleFilterChange} className="filter" />
      </div>
      <div className="stock">
        {
              filteredStock.map((item) => <StockItem key={stock.ticker} company={item} />)
          }
      </div>
    </div>
  );
};

const mapStateToProps = (state) => ({
  stock: state.stockReducer.stock,
  loading: state.stockReducer.loading,
  filter: state.filterReducer.category,
});
const mapDispatchToProps = (dispatch) => ({
  getStock: () => dispatch(getstock()),
});

StockCatalogue.propTypes = {
  getStock: PropTypes.func.isRequired,
  stock: PropTypes.instanceOf(Array).isRequired,
  loading: PropTypes.bool.isRequired,
  filter: PropTypes.string.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(StockCatalogue);
