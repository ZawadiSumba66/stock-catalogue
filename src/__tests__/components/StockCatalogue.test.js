import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';

import Adapter from 'enzyme-adapter-react-16';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import StockCatalogue from '../../components/StockCatalogue';
import store from '../../store';

configure({ adapter: new Adapter() });

it('matches the stock catalogue snapshot', () => {
  const tree = renderer.create(<Provider store={store}><StockCatalogue /></Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});

it('contains the <CategoryFilter/> component', () => {
  const component = shallow(<Provider store={store}><StockCatalogue /></Provider>);
  expect(component.find('CategoryFilter')).toBeTruthy();
});

it('renders loading before items are displayed', () => {
  const { getByTestId } = render(<Provider store={store}><StockCatalogue /></Provider>);
  expect(getByTestId('stock-catalogue')).toHaveTextContent('Please wait...');
});
