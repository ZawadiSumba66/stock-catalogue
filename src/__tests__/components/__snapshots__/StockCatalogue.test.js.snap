// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the stock catalogue snapshot 1`] = `
<div>
  <div
    className="header"
  >
    <span
      className="logo"
    >
      <i
        className="fas fa-poll"
      />
      STOCK MARKET
    </span>
    <form
      className="form"
      data-testid="category-filter"
    >
      Filter Stock:
       
      <select
        className="category-input"
        onChange={[Function]}
      >
        <option
          value="ALL"
        >
          ALL
        </option>
        <option
          value="Gainers"
        >
          Gainers
        </option>
        <option
          value="Losers"
        >
          Losers
        </option>
      </select>
    </form>
  </div>
  <div
    className="stock"
  />
</div>
`;
